sunpy/coordinates/utils.py

```diff
@@ -359,4 +373,13 @@ def get_rectangle_coordinates(bottom_left, *, top_right = None, width: u.deg = N
         if isinstance(bottom_left, BaseCoordinateFrame):
             top_right = top_right.frame
 
+    if top_right.spherical.lon < bottom_left.spherical.lon:
+        warnings.warn("The rectangle is inverted in the left/right (longitude) direction, possibly "
```

Comment:
Yes, it could be perfectly valid input, but it could also be user error.

Even if the user intends it, the reason for the warning is because downstream code may not interpret that rectangle in the intended fashion.  As currently implemented, both `submap()` and `draw_rectangle()` will not cross a longitude wrap that is at the edge of the map, and will instead incorrectly return the complement range in longitude.  Both might give the intended result for a longitude wrap that is in the middle of the map, but I haven't tested that.

Hence, that's why the warning says "...may lead to unintended behavior."

Reply:
Also see https://github.com/sunpy/sunpy/pull/3677#issuecomment-618417716

Reply:
But warnings shouldn't be raised for valid input that produces valid output, which can be the case here if the 0 deg longitude line is not on the edge of the map.

The answer might be not to sort the inputs in ascending pixel order, and then error if a user provides a `bottom_left` that isn't below and to the left of `top_right`, after the conversion has been done to pixel coordinates.

Reply:
I agree with not really wanting to raise warnings if the input & output are expected.

However, this also can't really rely on the conversion to pixel coords, as that only happens in submap not in `draw_rectangle` (where WCSAxes does the conversion).

Reply:
> But warnings shouldn't be raised for valid input that produces valid output

I disagree.  *Errors* certainly shouldn't be raised, but the warning here is a cautionary note.  Valid input may lead to valid output, but more likely than not will lead to undesired output.  Yes, this is mostly due to problematic code downstream, but that doesn't mean we shouldn't try to be helpful for the user.

Note that if the consensus is to remove this warning, then analogous warning/error code needs to be added to `submap()` and `draw_rectangle()` because they will both misbehave.  Having the warning here avoided having to add such code to those methods.

Reply:
It should be possible to check this though - something like `self.world_to_pixel(top_right).x < self.world_to_pixel(bottom_left).x` instead, would check if the rectangle is inverted.

Reply:
This is in a function without any WCS, not in a `Map` object.

Reply:
Then I guess the warning should move to `submap` so it has access? That way it would also be possible to conditionally raise the warning on heliographic maps and not helioprojective ones.

Reply:
I think we disagree about warning philosophy here.  I want to raise a warning if there's a reasonable chance that a user will encounter unintended behavior using the provided output.  You do not want to raise a warning if there's a reasonable chance that the user knows what he/she is doing.  I prefer my approach because a user can ignore/filter warnings.  Staying silent is not helpful to the user aside from just looking "clean".

Both `submap()` and `draw_rectangle()` can and probably should raise *errors*.  Even in that situation I think `get_rectangle_coordinates()` should still raise a warning.

Anyway, this is arguably largely beside the point for this PR, which is about consistent APIs.  The bugs that existed before this PR can still exist after this PR.  Let's not hold up this PR.

Reply:
My above comment was a solution to raise the warning if we *know* there will be unintended behaviour, I don't think it's a complicated or unreasonable change to make given this PR introduced this warning - I'm happy to make the change myself if it would help.

Reply:
For the sake of getting this merged, I'll put my proposed changes to the warning in a new PR once this one is merged.